### ISSUE TITLE
Plugin is not correctly generating variant/screen classes when using JIT

### ISSUE THREAD

**Issue by jjranalli** [tailwindlabs/tailwindcss#3972]
### What version of @tailwindcss/jit are you using?

v0.1.18

### What version of Node.js are you using?

v14.13.0

### What build tool (or framework if it abstracts the build tool) are you using?

postcss 8.2.9, Next.js 10.0.5, webpack 4.44.1

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### Reproduction repository

https://github.com/jjranalli/nightwind-jit-issue

The [Nightwind plugin](https://github.com/jjranalli/nightwind) essentially generates a bunch of classes like `.dark .text-red-600` to build a dark theme based on a specified tailwind.config.js configuration.

---
### Expected behavior
Similarly as when it's used without JIT, the plugin should generate classes for all TailwindCSS color classes – including variant/screen classes, or those added using `@apply`. For example, it should generate classes such as
- `.dark .test-hover:hover`
- `@media (min-width: 640px) .dark .test-sm-hover:hover`
- `@media (min-width: 640px) .dark .sm\:hover\:bg-rose-600:hover`

<img width="574" alt="Screenshot 2021-04-05 at 00 57 26" src="https://user-images.githubusercontent.com/39241410/113523755-26087700-95aa-11eb-8b86-fedb2789da5c.png">

[Reproduced here](https://nightwindcss-git-jit-comparison-jacoporanalli.vercel.app/)

---
### Outcome
When JIT is enabled, the same classes are not being generated.
 
<img width="574" alt="Screenshot 2021-04-05 at 00 58 53" src="https://user-images.githubusercontent.com/39241410/113523760-2e60b200-95aa-11eb-9747-7f2b77f78385.png">

[Reproduced minimally here](https://nightwind-jit-issue.vercel.app/), and with the full page [here](https://nightwindcss-git-jit-jacoporanalli.vercel.app/)

---

Since the bug is only present when JIT is enabled and only affects some variants and screen classes, I assume it has to do with the way it's generating those classes. 

In case this is not an issue on JIT side, I'd appreciate any feedback that could point me in the right direction so I can release a JIT-compatible version of the plugin. Thanks!

**Comment by adamwathan** [comment]
Hey sorry, can you create a very boiled down reproduction with a separate tiny plugin that isn't Nightwind? This is going to stay open forever otherwise as just not realistic that I'm going to find time to go deep into the Nightwind plugin and understand the entire codebase.

**Comment by jjranalli** [comment]
Thanks Adam, I understand. Unfortunately this is quite possibly related to how I built Nightwind, so I'm not sure I'd be able to reproduce it with a different plugin.

I've updated the [reproduction repo](https://github.com/jjranalli/nightwind-jit-issue) with a simplified version of Nightwind, inside tailwind.config.js. I've also updated tailwindcss to 2.1.2.

In a nutshell, this is what Nightwind does when generating color classes:

1. Creates an array with all combinations of color classes, including variants (i.e. ["text-red-600", "hover\\:text-red-600"])

```javascript
// const prefixes = ["text", "bg", "border"]
// const colors = theme("colors")
// const colorVariants = ["hover"]
// const weights = [50, 100, 200, 300, 400, 500, 600, 700, 800, 900]
prefixes.forEach((prefix) => {
  Object.keys(colors).forEach((color) => {
    weights.forEach((weight) => {
      let base = prefix + "-" + color + "-" + weight
      colorClasses.push(base)
      // Note: Here I'm generating variant classes iterating over variants provided in
      // tailwind.config.js. I did this because a '.dark' class will be prepended to all
      // classes at the next step, and I was having trouble generating the correct variants
      // with addComponents with the additional '.dark' class.
      colorVariants.forEach((variant) => {
        let baseVar = variant + "\\:" + prefix + "-" + color + "-" + weight
        colorClasses.push(baseVar)
      })
    })
  })
})
```

2. Iterates over the array to create an array of objects, containing the new dark-theme classes (based on the variant). For example:

```javascript
const nightwindClasses = [
  {
    [`.dark .hover\\:text-red-600:hover`]: {
      [`color`]: colorValue,
      [`color`]: hexToRGB(`${colorValue}`, `var(--tw-text-opacity)`),
    },
  },
  // ...
]
```

3. The classes are finally added using addComponents, including the responsive variants (which seems to be unnecessary with jit)

```javascript
addComponents(nightwindClasses, { variants: ["responsive"] })
```

One thing I should point out is that there’s some logic in the plugin that is not leveraging tailwindcss helper functions. Could this be the root cause of the issue? Is jit relying on them more during the compilation process?

The thing that bugs me is that everything works fine when jit is disabled (and has never caused problems before, even when applyComplexClasses was experimental), but when enabled it causes a range of issues that has to be related with one of the steps outlined above. For example, other than the issue related to variants, jit is also causing all nightwind classes to end up in production without being purged, resulting in a ~3-5x increase in css file size with respect to when it's disabled.

Unfortunately I don’t have the knowledge to understand what jit is doing under the hood (although I plan to look into it asap). So I'd greatly appreciate if you could share your thoughts about this, or if you could point me out the relevant parts of the codebase related to jit (especially those which are different from the past implementation and may be related to the issue), so I can try to figure this out on my own.

Thanks again!

**Comment by adamwathan** [comment]
Can you distill it down to a static example without a bunch of dynamic code? Like:

```js
addComponents({
  '.dark .hover\\:text-red-600:hover': {
    color: 'pink'
  },
}, { variants: ['responsive'] })
```

If that's the sort of stuff Nightwind ultimately inserts then we should be able to break it down to a very simple static example where you can just tell me what you expect to happen and what happens instead.

My gut feeling is it has to do with the fact that all of the variant modifiers are being baked in to the selectors, and I bet once I fully understand the problem here my recommendation is _probably_ going to be that Nightwind would be better as a PostCSS plugin rather than a Tailwind plugin, so you can just iterate over all of the user's CSS and generate dark versions of anything you detect that needs a dark version.

**Comment by jjranalli** [comment]
Thanks for the advice, will definitely consider turning Nightwind into a PostCSS plugin.

The generated rule is exactly as you wrote. I updated the repo again so that it only adds the relevant rules through addComponents, essentially removing the plugin. In the meantime I also think I finally get what's causing the issue.

As I said in my previous comment, in Nightwind I included logic to generate variant classes. The issue was that variants added through addComponents **target all classes** in the selector, including _".dark"_ or other additional classes generated by the plugin. This is why I only add _"responsive"_ to the variants in addComponents, instead of also including _"hover"_ or other variants.

By doing this, with JIT disabled, Nightwind correctly generates the following rules:

```js
// JIT disabled
addComponents(
  {
    ".dark .bg-rose-600": {
      backgroundColor: theme("colors.blue.300"),
    },
    ".dark .hover\\:bg-rose-600:hover": {
      backgroundColor: theme("colors.blue.300"),
    },
  },
  { variants: ["responsive"] }
)

// Respectevely generate:
// .dark .bg-rose-600 {...}
// .dark .hover\\:bg-rose-600:hover {...}
// + responsive media queries
```

However, when JIT is enabled variants are being generated by default and all classes are targeted. So when multiple classes are concatenated, this is what happens:

```js
// JIT enabled
addComponents({
  ".dark .bg-rose-600": {
    backgroundColor: theme("colors.blue.300"),
  },
  // Generates:
  // .dark .bg-rose-600 {...}
  // .hover\:dark:hover .hover\:bg-rose-600:hover {...}
  // + responsive classes

  ".dark .hover\\:bg-rose-600:hover": {
    backgroundColor: theme("colors.blue.300"),
  },
  // Generates:
  // .dark .hover\\:bg-rose-600:hover {...}
  // .hover\:dark:hover .hover\:bg-rose-600:hover {...} --> I assume it also generates this one, but it's duplicated
  // + responsive media queries
})
```

From my point of view, I see 3 possible solutions to the issue:

1. **Somehow prevent JIT from generating variants for classes added with addComponents**. This would allow me to generate the correct rules with minimal modifications to the existing plugin, but it would also prevent the plugin from getting JIT benefits (plugin variants would always needed to be added manually, for example). Also, I doubt this is even possible due to how JIT works.
2. **Prevent addComponents from targeting the _".dark"_ class and other additional classes when generating variants**. This would probably be the ideal solution, but it's not clear to me if I can achieve it through the functions provided by Tailwindcss.
3. As you already suggested, turning Nightwind into a PostCSS plugin.

In any case I look forward to hearing your thoughts, and thanks again for taking the time to go through all this! I really appreciate it
